### PR TITLE
Update year to 2022

### PR DESCRIPTION
**Pull request by hoffie**
**Short description of changes**

- [x] Automatical updates in translation source files:
`sed -re 's/-2021/-2022/g' -i src/res/translation/*.ts`

- [x] Manual verification/updates of all files found via:
`grep 2021 src/ *.cpp *.h -r |& grep -vP '\.(svg|qm)'`

- [ ] What do to about copyright code headers? I think one is supposed to update them when the file changes, but we currently do not enforce that. `src/main.cpp` still mentions 2020). This does not necessarily have to be done as part of this PR.

**Context: Fixes an issue?**

Thanks to @henkdegroot for the pointer. Release issue: #2247

**Does this change need documentation? What needs to be documented and how?**

No.

**Status of this Pull Request**

Ready.

**What is missing until this pull request can be merged?**

Review.

## Checklist

- [x] I've verified that this Pull Request follows the [general code principles](https://github.com/jamulussoftware/jamulus/blob/master/CONTRIBUTING.md#jamulus-projectsource-code-general-principles)
- [x] I tested my code and it does what I want
- [x] My code follows the [style guide](https://github.com/jamulussoftware/jamulus/blob/master/CONTRIBUTING.md#source-code-consistency) 
- [ ] I waited some time after this Pull Request was opened and all GitHub checks completed without errors. 
- [ ] I've filled all the content above

cc @softins

### CODE REVIEW

**Comment by pljones** [comment]
Ideally we'd have a check that any (source) file changed had the Copyright header correct - right latest year and the person making the change included in the list of authors for that file - that would fail the build, like clang-format verification.  Although it's always arguable whether that should be a build breaker.

**Comment by hoffie** [comment]
> Ideally we'd have a check that any (source) file changed had the Copyright header correct - right latest year and the person making the change included in the list of authors for that file - that would fail the build, like clang-format verification. Although it's always arguable whether that should be a build breaker.

Thinking some more, we could do that automatically based on git history as well. I will try to come up with something.

**Comment by hoffie** [comment]
> > Ideally we'd have a check that any (source) file changed had the Copyright header correct - right latest year and the person making the change included in the list of authors for that file - that would fail the build, like clang-format verification. Although it's always arguable whether that should be a build breaker.
> 
> Thinking some more, we could do that automatically based on git history as well. I will try to come up with something.

PR is updated and contains a small tool to update all known (?) global copyright strings, translation sources and copyright headers -- the latter based on `git log`.

The script is idem-potent, so it's safe to run anytime. We should still discuss where to run it, e.g.:
- a CI job on merges to master
- manually as part of the release checklist

**Comment by ann0see** [comment]
What about a scheduled job?

**Comment by hoffie** [comment]
> What about a schedueled job?

I'm currently looking into running this (and auto-committing) after any pushes to master (which hopefully includes Github merges?). The copyright header part is relevant during the whole year, so this would make sense. I'm testing in my private repos and will report back if I have something working.

File: linux/sound.cpp [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /******************************************************************************\
- * Copyright (c) 2004-2020
+ * Copyright (c) 2004-2021
```
Comment (ann0see) [inline linux/sound.cpp:2]: There’s something wrong here!

**Review by ann0see — CHANGES_REQUESTED** [review]
Seems inconsistent:

File: linux/sound.cpp (hunk 1 above)
Comment (hoffie) [reply linux/sound.cpp:2]: >  or would you really use the "last edited" date? I don’t see too much benefits from not just using the current year.

This was the intention and is what's currently implemented, yes.
Having read some more, the benefit of doing this 100% accurate seems low and the recommendation is to update all files consistenly.

I guess I should drop this logic again to simplify things.

https://www.gnu.org/prep/maintain/html_node/Copyright-Notices.html

**Comment by hoffie** [comment]
> What about a scheduled GH actions job? Best case would be that the job automatically opens a PR on 01-01-xxxx

Yes, that should work, there's a `schedule` event for workflows. I still tend to trigger on pushes to master instead:
- The year should only be updated if there are significant changes. Significant changes should always come via a push.
- A PR might introduce new code with wrong copyright dates. Therefore, it makes sense to run after each PR.

We should still check if we want to auto-open a PR (a little bit more effort) or simply auto-fix via a commit by CI (simple, but less transparent).

**Comment by ann0see** [comment]
Doing it via commit once a year might result in unintended conflicts. I‘d definitely be in favour of a PR.

**Comment by hoffie** [comment]
> Doing it via commit once a year might result in unintended conflicts. I‘d definitely be in favour of a PR.

I prefer a PR for transparency as well, especially considering that this should only happen a few times (maybe only once) a year. I have just updated the code and the PR description accordingly.
Removing Draft status.

File: linux/sound.cpp (hunk 1 above)
Comment (hoffie) [reply linux/sound.cpp]: The updated logic no longer inserts file-specific dates, it always updates to the current year. Resolving.

File: tools/update-copyright-notices.sh [hunk 2]
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
+set -eu
+
+YEAR=$(date +%Y)
+echo "Updating global copyright strings..."
+sed -re 's/(Copyright.*20[0-9]{2}-)[0-9]{4}/\1'$YEAR'/g' -i src/res/translation/*.ts src/util.cpp src/aboutdlgbase.ui
```
Comment (ann0see) [inline tools/update-copyright-notices.sh:6]: Although a bit picky (and we still have a bunch of years left): but could you use a less specific regex in terms of not specifying the prefix 20 for the year?

File: .github/workflows/update-copyright-notices.yml [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+jobs:
+  check-and-update-copyright-notices:
+    if: ${{ github.event_name == 'push' }}
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+      - run: ./tools/update-copyright-notices.sh
+      - run: |
+          git config --global user.email "actions@github.com"
```
Comment (ann0see) [inline .github/workflows/update-copyright-notices.yml:19]: Does this need to be set explicitly?

File: .github/workflows/update-copyright-notices.yml [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+          git config --global user.name "github-actions[bot]"
+          pr_branch=update-copyright-notices-$(date +%Y-%m-%d-%H-%M-%S)
```
Comment (ann0see) [inline .github/workflows/update-copyright-notices.yml:21]: I‘d prefer something shorter like `updateCopyrightYear/<year>`

File: .github/workflows/update-copyright-notices.yml [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+          git checkout -b ${pr_branch}
+          git add .
+          git commit -m "Update copyright notices for $(date +%Y)" || exit 0
```
Comment (ann0see) [inline .github/workflows/update-copyright-notices.yml:24]:
```suggestion
          git commit -m "AUTO: Update copyright notices for $(date +%Y)" || exit 0
```

Similar to what the website does.

**Review by ann0see — COMMENTED** [review]
Thanks for all the automation stuff you’ve doing ;-).

**Comment by ann0see** [comment]
@hoffie 

could the PR say something like:

This automated Pull Request updates the copyright messages throughout the source.

Output of […]

File: .github/workflows/update-copyright-notices.yml (hunk 3 above)
Comment (hoffie) [reply .github/workflows/update-copyright-notices.yml:19]: Yes. The defaults error out and passing `git commit --author` is not sufficient either.

**Comment by ann0see** [comment]
Although unrelated, the contributor checking script could be automated the same way

File: tools/update-copyright-notices.sh (hunk 2 above)
Comment (hoffie) [reply tools/update-copyright-notices.sh:6]: Done, although I've kept the `2`. I really want to avoid that the script touches an unintended place with its modifications.

File: .github/workflows/update-copyright-notices.yml (hunk 4 above)
Comment (hoffie) [reply .github/workflows/update-copyright-notices.yml:21]: Thinking some more, we can even drop the year altogether. There should only ever be one active PR. If a new one is needed, the existing one can be updated (= the branch can be force-pushed). I've updated this accordingly.

File: .github/workflows/update-copyright-notices.yml (hunk 5 above)
Comment (hoffie) [reply .github/workflows/update-copyright-notices.yml:24]: Done (although the fact that it's automated should already be clear by looking at the author).

File: tools/update-copyright-notices.sh (hunk 2 above)
Comment (ann0see) [reply tools/update-copyright-notices.sh:6]: Understandable 😀.

File: .github/workflows/update-copyright-notices.yml (hunk 5 above)
Comment (ann0see) [reply .github/workflows/update-copyright-notices.yml:24]: Yes. Just a consistency thing.

File: .github/workflows/update-copyright-notices.yml [hunk 6]
```diff
@@ -0,0 +1,42 @@
+name: Update copyright notices
+
+on:
+  push:
+    branches:
+      - master
+  pull_request:
```
Comment (ann0see) [inline .github/workflows/update-copyright-notices.yml:7]: Do we need to run this on PR close with the simplified logic?
I‘d prefer to run it once/twice a year. Preferably in January.

Comment (hoffie) [reply .github/workflows/update-copyright-notices.yml:7]: Not sure, I'd still run it on every push/merge because each PR may introduce old headers again.

Indepdendent of the frequency we can drop the deletion logic if we are OK to permanently have an updateChangelogNotices branch in jamulussoftware/jamulus.

Comment (ann0see) [reply .github/workflows/update-copyright-notices.yml:7]: Ok. Got your concern. I wouldn’t drop the deletion logic since it might clutter the repo. Maybe we‘d even need to ensure different names of the branch since conflicting history on a local branch and on GH might confuse people pulling the repo.

**Comment by hoffie** [comment]
> could the PR say something like:
> 
> This automated Pull Request updates the copyright messages throughout the source.
> 
> Output of […]

Done. Examples above have been updated to show the updated texts/behavior.

> Although unrelated, the contributor checking script could be automated the same way

Yes. Can you open a seperate issue?

File: .github/workflows/update-copyright-notices.yml (hunk 6 above)
Comment (hoffie) [reply .github/workflows/update-copyright-notices.yml:7]: Heh, that was the initial reason for the rather specific branch names. However, having those would clutter people's checkouts (at least when they check them out). I'd keep it like it is now (single branch name) as noone should usually check out those automatic repos. If they do, they'd hopefully know why. :)

Comment (ann0see) [reply .github/workflows/update-copyright-notices.yml:7]: Yes. Great!

File: .github/workflows/update-copyright-notices.yml [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+          git checkout -b ${pr_branch}
+          git add .
+          git commit -m "AUTO: Update copyright notices for $(date +%Y)" || exit 0
+          git push origin "+${pr_branch}"
+          existing_pr=$(gh pr list --head "${pr_branch}" --json number --jq '.[].number')
+          if [[ $existing_pr ]]; then
+            gh pr comment "${existing_pr}" --body "PR has been updated again."
+            exit 0
+          fi
+          gh pr create --base test-master --head "${pr_branch}" --title "Update copyright notice(s) for $(date +%Y)" --body $'This automated Pull Request updates the copyright notices throughout the source.\n\nThis PR is the result of a `./tools/update-copyright-notices.sh` run.'
```
Comment (ann0see) [inline .github/workflows/update-copyright-notices.yml:33]: You'll probably change test-master to master once this is finished?

**Review by ann0see — APPROVED** [review]

File: .github/workflows/update-copyright-notices.yml [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+          git config --global user.name "github-actions[bot]"
+          pr_branch=updateCopyrightNotices
+          git checkout -b ${pr_branch}
+          git add .
+          git commit -m "AUTO: Update copyright notices for $(date +%Y)" || exit 0
+          git push origin "+${pr_branch}"
+          existing_pr=$(gh pr list --head "${pr_branch}" --json number --jq '.[].number')
+          if [[ $existing_pr ]]; then
+            gh pr comment "${existing_pr}" --body "PR has been updated again."
+            exit 0
```
Comment (ann0see) [inline .github/workflows/update-copyright-notices.yml:31]: I'd prefer a if/else instead of exiting here (this makes the code clearer)

File: .github/workflows/update-copyright-notices.yml [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+          git config --global user.email "actions@github.com"
+          git config --global user.name "github-actions[bot]"
+          pr_branch=updateCopyrightNotices
+          git checkout -b ${pr_branch}
+          git add .
+          git commit -m "AUTO: Update copyright notices for $(date +%Y)" || exit 0
+          git push origin "+${pr_branch}"
+          existing_pr=$(gh pr list --head "${pr_branch}" --json number --jq '.[].number')
+          if [[ $existing_pr ]]; then
+            gh pr comment "${existing_pr}" --body "PR has been updated again."
```
Comment (ann0see) [inline .github/workflows/update-copyright-notices.yml:30]: @gilgongo not sure if this message is clear enough?

File: .github/workflows/update-copyright-notices.yml (hunk 8 above)
Comment (hoffie) [reply .github/workflows/update-copyright-notices.yml]: Done.

File: .github/workflows/update-copyright-notices.yml (hunk 7 above)
Comment (hoffie) [reply .github/workflows/update-copyright-notices.yml]: Umm, I had missed that. Thanks! Fixed.

File: .github/workflows/update-copyright-notices.yml (hunk 9 above)
Comment (hoffie) [reply .github/workflows/update-copyright-notices.yml:30]: What should be changes/added?

Comment (ann0see) [reply .github/workflows/update-copyright-notices.yml:30]: We should clarify *why* the PR was updated. Is it possible to link the related commit? Then something like:

Updated this PR because of commit xyz
At least for me it wasn't quite clear why this message would come up.

Comment (hoffie) [reply .github/workflows/update-copyright-notices.yml:30]: > We should clarify why the PR was updated.

Is it really relevant? The updated diff is easily visible from the PR. We don't explain the *why* in more depth when initially creating the PR either. The answer is simple: The script becomes active (opens a PR / updates it) when there's something to fix. This might be due to the year rolling over or due to some old copyright header being introduced.

Looking up the commit which triggered is probably possible in a simple way. But remember that this is push-based, so it will not be easy to tell if it's this commit or rather some parent.

To sum it up: It would certainly be possible, but it would lead to way more code/complexity.

Comment (ann0see) [reply .github/workflows/update-copyright-notices.yml:30]: Ok. So you'd say it's not worth it. I'm still not happy with a message saying (Yeah, this PR just got updated). What speaks against omitting it completely? What speaks for having it? I mean, you'll always see on GH that the PR got force-pushed and hence updated.

Comment (hoffie) [reply .github/workflows/update-copyright-notices.yml:30]: Well, it's an edge case (IMO).
I was mostly thinking about protecting against unwanted merges.
E.g.:
- PR is reviewed and close to being merged
- Script updates PR
- PR is merged, thinking that it had been reviewed, without anyone noticing that it was changed meanwhile

This can happen with human PRs as well, but humans might act in good faith and provide a pointer, so I figured the script should try do the same. :)

Comment (pljones) [reply .github/workflows/update-copyright-notices.yml:30]: In which case 
> "PR has been updated again."

could be
> "PR was updated by tools/update-copyright-headers"

or similar, to make it clear.

Comment (hoffie) [reply .github/workflows/update-copyright-notices.yml]: Done.

File: .github/workflows/update-copyright-notices.yml (hunk 8 above)
Comment (softins) [reply .github/workflows/update-copyright-notices.yml]: Personally, I am a big fan of early exit/return, and find it much clearer.

If there is nothing further to do, why not exit at that point, rather than have the reader need to follow through the else/fi to find there is nothing else below?

Comment (hoffie) [reply .github/workflows/update-copyright-notices.yml]: > Personally, I am a big fan of early exit/return, and find it much clearer.

Same here. I guess this instance is less clear than others though as there are two code paths where neither is the clear default. If the script was to be extended, I couldn't predict which of those two paths it would be.

File: .github/workflows/update-copyright-notices.yml (hunk 9 above)
Comment (ann0see) [reply .github/workflows/update-copyright-notices.yml]: Ok. Probably it’s ok then.

File: .github/workflows/update-copyright-notices.yml [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+          pr_branch=updateCopyrightNotices
+          git checkout -b ${pr_branch}
+          git add .
+          git commit -m "AUTO: Update copyright notices for $(date +%Y)" || exit 0
+          git push origin "+${pr_branch}"
+          existing_pr=$(gh pr list --head "${pr_branch}" --json number --jq '.[].number')
+          if [[ $existing_pr ]]; then
+            gh pr comment "${existing_pr}" --body "PR has been updated by tools/update-copyright-notices.sh."
+          else
+            gh pr create --base master --head "${pr_branch}" --title "Update copyright notice(s) for $(date +%Y)" --body $'This automated Pull Request updates the copyright notices throughout the source.\n\nThis PR is the result of a `tools/update-copyright-notices.sh` run.'
```
Comment (pljones) [inline .github/workflows/update-copyright-notices.yml:32]: Could you `\` wrap this line, please.  (We don't have clang-format for YAML :) )

Comment (hoffie) [reply .github/workflows/update-copyright-notices.yml]: Hrm, even when wrapping the text stays rather long. I've therefore moved the text to a variable and build that in two lines. Should be better now.

File: tools/update-copyright-notices.sh [hunk 11]
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
+set -eu
+
+YEAR=$(date +%Y)
+echo "Updating global copyright strings..."
+sed -re 's/(Copyright.*2[0-9]{3}-)[0-9]{4}/\1'$YEAR'/g' -i src/res/translation/*.ts src/util.cpp src/aboutdlgbase.ui
+
+echo "Updating copyright comment headers..."
+for file in $(find -regex '.*\.\(cpp\|h\|mm\)' -not -regex '\./\(\.git\|libs/\|moc_\|ui_\).*'); do
```
Comment (softins) [inline tools/update-copyright-notices.sh:9]: I just tried this out and found it had the same issue as the recent clang_format one: it found all the files in build directories too.
```suggestion
for file in $(find android ios linux mac src windows -regex '.*\.\(cpp\|h\|mm\)' -not -regex '\./\(\.git\|libs/\|moc_\|ui_\).*'); do
```

Comment (hoffie) [reply tools/update-copyright-notices.sh]: Thansk, updated.

**Review by pljones — APPROVED** [review]

**Review by softins — APPROVED** [review]
